### PR TITLE
interactive parametrized SQL queries

### PR DESCRIPTION
**Pull request by tonykploomber**
## Issue ticket number and link
Closes #x

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [ ] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--293.org.readthedocs.build/en/293/

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -18,14 +18,15 @@
...
-    "sqlalchemy<2",
+    "sqlalchemy<=1.4.46",
     "sqlparse",
     "ipython-genutils>=0.1.0",
     "sqlglot",
     "jinja2",
     "sqlglot>=11.3.7",
     "ploomber-core>=0.2.4",
-    'importlib-metadata;python_version<"3.8"'
+    "ipywidgets",
```
Comment (edublancas) [inline setup.py:28]: this will make ipywidgets a required dependency, let's make it optional by calling [`check_installed`](https://ploomber-core.readthedocs.io/en/latest/dependencies.html#check-installed) if we detect the user passed `--interactive`

Let's include this in the tutorial so they know they need it, something like "ipywidget is required to use this feature, install it with pip install ..."

File: src/sql/command.py [hunk 2]
```diff
@@ -88,4 +87,22 @@ def result_var(self):
...
+            "Variable substitution with $var and {var} has been "
+            "deprecated and will be removed in a future version. "
+            "Use {{var}} instead. To remove this, see: "
+            "https://jupysql.ploomber.io/en/latest/howto.html#ignore-deprecation-warnings"  # noqa
+        )
+
+        if parsed_sql != sql or has_SQLAlchemy_var_expand:
+            warnings.warn(msg, FutureWarning)
+
+        return parsed_sql
```
Comment (edublancas) [inline src/sql/command.py:108]: this looks unrelated to the PR

Comment (edublancas) [reply src/sql/command.py:108]: maybe added by mistake?

File: src/sql/magic.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import json
 import re
+from ipywidgets import interact
```
Comment (edublancas) [inline src/sql/magic.py:3]: let's make this an optional dependency

```suggestion
try:
    from ipywidgets import interact
except ModuleNotFoundError:
    interact = None
```

File: src/sql/magic.py [hunk 4]
```diff
@@ -264,6 +278,17 @@ def _execute(self, payload, line, cell, local_ns):
         # args.line: contains the line after the magic with all options removed
 
         args = command.args
+        # Create the interactive slider
```
Comment (edublancas) [inline src/sql/magic.py:281]: we need to add a check here, if the user passed the interact argument, then we check that ipywidgets is installed using `check_installed`

File: doc/user-guide/interactive.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.14.5
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# Interactive SQL Queries
```
Comment (edublancas) [inline doc/user-guide/interactive.md:14]:
~~~suggestion
# Interactive SQL Queries

```{versionadded} 0.7
```

~~~

**Review by edublancas — CHANGES_REQUESTED** [review]

**Comment by edublancas** [comment]
also, please address the merging conflicts

File: setup.py (hunk 1 above)
Comment (edublancas) [reply setup.py:28]: note that I just added `check_installed` to ploomber core so we need to pin it in setup.py `ploomber-core>=0.2.7`

File: CHANGELOG.md [hunk 6]
```diff
@@ -5,6 +5,7 @@
 * [API Change] Deprecates old SQL parametrization: `$var`, `:var`, and `{var}` in favor of `{{var}}`
 * [Fix] `--save` + `--with` double quotes syntax error in MySQL ([#145](https://github.com/ploomber/jupysql/issues/145))
 * [Feature] Adds sql magic test to list of possible magics to test datasets
+* [Feature] Adds `--interactive` with ipywidgets Widget (#256)
```
[the remark this answers is not in the document]
Comment (tonykploomber) [reply CHANGELOG.md]: Fix in [82184f6](https://github.com/ploomber/jupysql/pull/293/commits/82184f6bdc7731a290463ff834731efb4d1af651)

File: doc/user-guide/interactive.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+Interactive command allows you to visualize and manipulate widget and interact with your SQL cluase.
+We will demonstrate how to create widgets and dynamically query the dataset.
+
+```{note}
+`%sql --interactive` requires `ipywidgets`: `pip install ipywidgets`
+```
+
+## `%sql --interacrive {{widget_variable}}`
```
[the remark this answers is not in the document]
Comment (tonykploomber) [reply doc/user-guide/interactive.md]: Fix in [82184f6](https://github.com/ploomber/jupysql/pull/293/commits/82184f6bdc7731a290463ff834731efb4d1af651)

File: src/sql/command.py (hunk 2 above)
Comment (tonykploomber) [reply src/sql/command.py]: You right, some merge error
FIx in [a619224](https://github.com/ploomber/jupysql/pull/293/commits/a619224681751aa113b43f073828002f6f2499cd)

File: src/sql/magic.py (hunk 3 above)
Comment (tonykploomber) [reply src/sql/magic.py]: Fix in [82184f6](https://github.com/ploomber/jupysql/pull/293/commits/82184f6bdc7731a290463ff834731efb4d1af651)

File: doc/user-guide/interactive.md (hunk 5 above)
Comment (tonykploomber) [reply doc/user-guide/interactive.md:14]: Fix in [82184f6](https://github.com/ploomber/jupysql/pull/293/commits/82184f6bdc7731a290463ff834731efb4d1af651)

File: src/sql/magic.py (hunk 4 above)
Comment (tonykploomber) [reply src/sql/magic.py:285]: Handled in [e6b45dd](https://github.com/ploomber/jupysql/pull/293/commits/e6b45dd39b8b01856b8351407f346fe2ad583668)

File: setup.py (hunk 1 above)
Comment (tonykploomber) [reply setup.py:28]: Handled in [e6b45dd](https://github.com/ploomber/jupysql/pull/293/commits/e6b45dd39b8b01856b8351407f346fe2ad583668)

**Review by edublancas — APPROVED** [review]
